subsys/bluetooth/services/ddfs.c

```diff
@@ -0,0 +1,419 @@
+{
+	ddfs_inst.dm_ctrl_pt_indicate_enabled = (value == BT_GATT_CCC_INDICATE);
+}
+
+BT_GATT_SERVICE_DEFINE(ddf_svc,
+	BT_GATT_PRIMARY_SERVICE(BT_UUID_DDFS),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DDFS_DISTANCE_MEAS, BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_NONE, NULL, NULL, NULL),
+	BT_GATT_CCC(ddf_distance_meas_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DFFS_AZIMUTH_MEAS, BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_NONE, NULL, NULL, NULL),
+	BT_GATT_CCC(ddf_azimuth_meas_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DFFS_ELEVATION_MEAS, BT_GATT_CHRC_NOTIFY,
+			       BT_GATT_PERM_NONE, NULL, NULL, NULL),
+	BT_GATT_CCC(ddf_elevation_meas_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DDFS_FEATURE, BT_GATT_CHRC_READ,
+			       BT_GATT_PERM_READ, read_ddf_dm_featute, NULL, NULL),
+	BT_GATT_CHARACTERISTIC(BT_UUID_DFFS_CTRL_POINT, BT_GATT_CHRC_WRITE |  BT_GATT_CHRC_INDICATE,
+			       BT_GATT_PERM_WRITE, NULL, dm_write_ctrl_point, NULL),
+	BT_GATT_CCC(dm_ctrl_point_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+);
+
+static void ctrl_point_ind(struct bt_conn *conn, uint8_t req_op, uint8_t status,
+			   const void *data, uint16_t len)
+{
+	struct dm_ctrl_point_ind *ind;
+
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(*ind) + len);
+
+	if (ddfs_inst.indicating) {
+		return;
+	}
+
+	ind = net_buf_simple_add(&buf, sizeof(*ind));
+	ind->op = DDF_OP_RESPONSE;
+	ind->req_op = req_op;
+	ind->status = status;
+
+	if (data && len) {
+		net_buf_simple_add_mem(&buf, data, len);
+	}
+
+	ddfs_inst.ind_params.attr = &ddf_svc.attrs[DDF_SVC_CTRL_POINT_ATTR_IDX];
+	ddfs_inst.ind_params.func = indicate_cb;
+	ddfs_inst.ind_params.destroy = indicate_destroy;
+	ddfs_inst.ind_params.data = ind;
+	ddfs_inst.ind_params.len = buf.len;
+
+	if (bt_gatt_indicate(conn, &ddfs_inst.ind_params) == 0) {
+		ddfs_inst.indicating = true;
+	}
+}
+
+int bt_ddfs_distance_measurement_notify(struct bt_conn *conn,
+					const struct bt_ddfs_distance_measurement *measurement)
+{
+	if (!ddfs_inst.dm_notification_enabled) {
+		return -EACCES;
+	}
+
+	if (!measurement) {
+		return -EINVAL;
+	}
+
+	bool is_rtt = measurement->ranging_mode == BT_DDFS_DM_RANGING_MODE_RTT;
+	bool is_mcpd = measurement->ranging_mode == BT_DDFS_DM_RANGING_MODE_MCPD;
+	struct ddfs_meas_notify *dist_meas_notify;
+
+	NET_BUF_SIMPLE_DEFINE(buf, sizeof(*dist_meas_notify) +
+				   (is_rtt ? sizeof(measurement->dist_estimates.rtt) : 0) +
+				   (is_mcpd ? sizeof(measurement->dist_estimates.mcpd) : 0));
+
+	dist_meas_notify = net_buf_simple_add(&buf, sizeof(*dist_meas_notify));
+	dist_meas_notify->flags = 0;
+	dist_meas_notify->quality = measurement->quality;
+	dist_meas_notify->bt_addr[6] = measurement->bt_addr.type;
+
+	for (size_t i = 0; i < 6; i++) {
+		dist_meas_notify->bt_addr[i] = measurement->bt_addr.a.val[i];
```

Comment:
Maybe just `memcpy` ?
The same in other places

Reply:
Good point.